File: VB/PLinqServerModeSource/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Skins
Imports DevExpress.LookAndFeel
Imports DevExpress.UserSkins
Imports DevExpress.XtraEditors
Imports DevExpress.Data.PLinq
Imports System.Linq
Imports System.Data.Linq

Namespace PLinqServerMode
	Partial Public Class Form1
		Inherits XtraForm
		Public Sub New()
			InitializeComponent()
			AddHandler plIFS.GetEnumerable, AddressOf plIFS_GetEnumerable
			AddHandler gridView1.AsyncCompleted, AddressOf gridView1_AsyncCompleted
			oldRowsCount = gridView1.DataRowCount
			gridControl.DataSource = plIFS
		End Sub

		Private Sub plIFS_GetEnumerable(ByVal sender As Object, ByVal e As GetEnumerableEventArgs)
			e.Source = nwdContext.Customers
		End Sub

		Private Sub gridView1_AsyncCompleted(ByVal sender As Object, ByVal e As EventArgs)
			If customerToEdit IsNot Nothing AndAlso gridView1.DataRowCount > oldRowsCount Then
				For i As Integer = 0 To gridView1.DataRowCount - 1
					If customerToEdit.CustomerID = gridView1.GetRowCellValue(i, gridView1.Columns("CustomerID")).ToString() Then
						gridView1.FocusedRowHandle = i
						oldRowsCount = gridView1.DataRowCount
						Exit For
					End If
				Next i
			End If
		End Sub
		Private plIFS As New PLinqInstantFeedbackSource()
		Private nwdContext As New NorthwindDataContext()
		Private customerToEdit As Customer
		Private oldRowsCount As Integer = 0
		Private f1 As EditForm

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			customerToEdit = CreateCustomer()
			oldRowsCount = gridView1.DataRowCount
			EditCustomer(customerToEdit, "NewCustomer", AddressOf CloseNewCustomerHandler)
		End Sub
		Private Function CreateCustomer() As Customer
			Dim idString As String
			Dim newCustomer = New Customer()
			Do
				idString = GenerateCustomerID()
				If (Not String.IsNullOrEmpty(idString)) Then
					newCustomer.CustomerID = idString
					Exit Do
				End If
			Loop
			Return newCustomer
		End Function

		Private Function GenerateCustomerID() As String
			Const IDLength As Integer = 5
			Dim result = String.Empty
			Dim rnd = New Random()
			Dim collisionFlag As Boolean = False
			For i = 0 To IDLength - 1
				result += Convert.ToChar(rnd.Next(65, 90))
			Next i
			For i As Integer = 0 To gridView1.DataRowCount - 1
				If result Is gridView1.GetRowCellValue(i, gridView1.Columns("CustomerID")).ToString() Then
					collisionFlag = True
					Exit For
				End If
			Next i
			If collisionFlag Then
				Return String.Empty
			Else
				Return result
			End If
		End Function
		Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton2.Click
            Dim query = From
                                                Customer In
                                                nwdContext.Customers
                                                         Where
                                                         Customer.CustomerID = GetCustomerIDByRowHandle(gridView1.FocusedRowHandle)
                                                         Select
                                                         Customer
			customerToEdit = query.ToList()(0)
			EditCustomer(customerToEdit, "EditInfo", AddressOf CloseEditCustomerHandler)
		End Sub
		Private Sub EditCustomer(ByVal customer As Customer, ByVal windowTitle As String, ByVal closedDelegate As FormClosingEventHandler)
			f1 = New EditForm(customer) With {.Text = windowTitle}
			AddHandler f1.FormClosing, closedDelegate
			f1.ShowDialog()
		End Sub
		Private Function GetCustomerIDByRowHandle(ByVal rowHandle As Integer) As String
			Return CStr(gridView1.GetRowCellValue(rowHandle, "CustomerID"))
		End Function
		Private Sub CloseEditCustomerHandler(ByVal sender As Object, ByVal e As EventArgs)
			If (CType(sender, EditForm)).DialogResult = System.Windows.Forms.DialogResult.OK Then
				Try
					nwdContext.SubmitChanges()
					nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers)
					plIFS.Refresh()

				Catch ex As Exception
					HandleExcepton(ex)
				End Try
			End If
			customerToEdit = Nothing
		End Sub
		Private Sub CloseNewCustomerHandler(ByVal sender As Object, ByVal e As FormClosingEventArgs)
			If (CType(sender, EditForm)).DialogResult = System.Windows.Forms.DialogResult.OK Then
				Try
					nwdContext.Customers.InsertOnSubmit(customerToEdit)
					nwdContext.SubmitChanges()
					nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers)
					plIFS.Refresh()

				Catch ex As Exception
					HandleExcepton(ex)
				End Try
			Else
				customerToEdit = Nothing
			End If
		End Sub

		Private Sub HandleExcepton(ByVal ex As Exception)
			MessageBox.Show(ex.Message)
		End Sub

		Private Sub button3_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton3.Click
			DeleteCustomer(gridView1.FocusedRowHandle)
		End Sub
		Private Sub DeleteCustomer(ByVal focusedRowHandle As Integer)
			If focusedRowHandle < 0 Then
				Return
			End If
			If MessageBox.Show("Do you really want to delete the selected customer?", "Delete Customer", MessageBoxButtons.OKCancel) <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If
            Dim query = From
                                                         Customer In
                                                         nwdContext.Customers
                                                                  Where
                                                                  Customer.CustomerID = GetCustomerIDByRowHandle(gridView1.FocusedRowHandle)
                                                                  Select
                                                                  Customer
			nwdContext.Customers.DeleteOnSubmit(query.ToList()(0))
			Try
				nwdContext.SubmitChanges()
				nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers)
			Catch ex As Exception
				HandleExcepton(ex)
			End Try
			plIFS.Refresh()
			customerToEdit = Nothing
		End Sub
	End Class
End Namespace

File: CS/PLinqServerModeSource/Form1.cs
// Developer Express Code Central Example:
// How to implement CRUD operations using XtraGrid and PLinqInstantFeedbackSource
// 
// This example demonstrates how to implement the Create, Update and Delete
// operations using PLinqInstantFeedbackSource.
// This example works with the
// standard SQL Northwind database.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4501

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.LookAndFeel;
using DevExpress.UserSkins;
using DevExpress.XtraEditors;
using DevExpress.Data.PLinq;
using System.Linq;
using System.Data.Linq;

namespace PLinqServerMode
{
    public partial class Form1 : XtraForm
    {
        public Form1()
        {
            InitializeComponent();           
            plIFS.GetEnumerable += new EventHandler<GetEnumerableEventArgs>(plIFS_GetEnumerable);
            gridView1.AsyncCompleted += new EventHandler(gridView1_AsyncCompleted);
            oldRowsCount = gridView1.DataRowCount;
            gridControl.DataSource = plIFS;
        }

        void plIFS_GetEnumerable(object sender, GetEnumerableEventArgs e)
        {
            e.Source = nwdContext.Customers;
        }

        void gridView1_AsyncCompleted(object sender, EventArgs e)
        {
            if (customerToEdit != null && gridView1.DataRowCount > oldRowsCount)
            {
                for (int i = 0; i < gridView1.DataRowCount; i++)
                {
                    if (customerToEdit.CustomerID == gridView1.GetRowCellValue(i, gridView1.Columns["CustomerID"]).ToString())
                    {
                        gridView1.FocusedRowHandle = i;
                        oldRowsCount = gridView1.DataRowCount;
                        break;
                    }
                }
            }
        }
        PLinqInstantFeedbackSource plIFS = new PLinqInstantFeedbackSource();
        NorthwindDataContext nwdContext = new NorthwindDataContext();
        Customer customerToEdit;
        int oldRowsCount = 0;
        EditForm f1;

        private void button1_Click(object sender, EventArgs e)
        {
            customerToEdit = CreateCustomer();
            oldRowsCount = gridView1.DataRowCount;
            EditCustomer(customerToEdit, "NewCustomer", CloseNewCustomerHandler);
        }
        private Customer CreateCustomer()
        {
            string idString;
            var newCustomer = new Customer();
            while (true)
            {
                idString = GenerateCustomerID();
                if (!String.IsNullOrEmpty(idString))
                {
                    newCustomer.CustomerID = idString;
                    break;
                }
            }
            return newCustomer;
        }

        private string GenerateCustomerID()
        {
            const int IDLength = 5;
            var result = String.Empty;
            var rnd = new Random();
            bool collisionFlag = false;
            for (var i = 0; i < IDLength; i++)
            {
                result += Convert.ToChar(rnd.Next(65, 90));
            }
            for (int i = 0; i < gridView1.DataRowCount; i++)
            {
                if (result == gridView1.GetRowCellValue(i, gridView1.Columns["CustomerID"]).ToString())
                {
                    collisionFlag = true;
                    break;
                }
            }
            if (collisionFlag)
            {
                return String.Empty;
            }
            else
                return result;
        }
        private void button2_Click(object sender, EventArgs e)
        {
            var query = nwdContext.Customers.Where<Customer>(customer => customer.CustomerID == GetCustomerIDByRowHandle(gridView1.FocusedRowHandle));
            customerToEdit = query.ToList()[0];
            EditCustomer(customerToEdit, "EditInfo", CloseEditCustomerHandler);
        }
        private void EditCustomer(Customer customer, string windowTitle, FormClosingEventHandler closedDelegate)
        {
            f1 = new EditForm(customer) { Text = windowTitle };
            f1.FormClosing += closedDelegate;
            f1.ShowDialog();
        }
        private string GetCustomerIDByRowHandle(int rowHandle)
        {
            return (string)gridView1.GetRowCellValue(rowHandle, "CustomerID");
        }
        private void CloseEditCustomerHandler(object sender, EventArgs e)
        {
            if (((EditForm)sender).DialogResult == DialogResult.OK)
            {
                try
                {
                    nwdContext.SubmitChanges();
                    nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers);
                    plIFS.Refresh();

                }
                catch (Exception ex)
                {
                    HandleExcepton(ex);
                }
            }
            customerToEdit = null;
        }
        private void CloseNewCustomerHandler(object sender, FormClosingEventArgs e)
        {
            if (((EditForm)sender).DialogResult == DialogResult.OK)
            {
                try
                {
                    nwdContext.Customers.InsertOnSubmit(customerToEdit);
                    nwdContext.SubmitChanges();
                    nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers);
                    plIFS.Refresh();

                }
                catch (Exception ex)
                {
                    HandleExcepton(ex);
                }
            }
            else
            {
                customerToEdit = null;
            }
        }

        private void HandleExcepton(Exception ex)
        {
            MessageBox.Show(ex.Message);
        }

        private void button3_Click(object sender, EventArgs e)
        {
            DeleteCustomer(gridView1.FocusedRowHandle);
        }
        private void DeleteCustomer(int focusedRowHandle)
        {
            if (focusedRowHandle < 0)
                return;
            if (MessageBox.Show("Do you really want to delete the selected customer?", "Delete Customer", MessageBoxButtons.OKCancel) != DialogResult.OK)
            {
                return;
            }
            var query = nwdContext.Customers.Where<Customer>(customer => customer.CustomerID == GetCustomerIDByRowHandle(focusedRowHandle));
            nwdContext.Customers.DeleteOnSubmit(query.ToList()[0]);
            try
            {
                nwdContext.SubmitChanges();
                nwdContext.Refresh(RefreshMode.OverwriteCurrentValues, nwdContext.Customers);
            }
            catch (Exception ex)
            {
                HandleExcepton(ex);
            }
            plIFS.Refresh();
            customerToEdit = null;
        }
    } 
}